Add register endpoint

jamesmaa opened the pull request:
👓 @schwers @nramadas 

How does QA work? Do I write tests or push to staging somewhere?

Commit:
Add register endpoint

nramadas commented on lib/api.es6.js:
```diff
@@ -50,6 +50,85 @@ const login = (apiOptions, username, pass) => {
+      data.email = email;
+    }
+
+    if (newsletter.length === 1) {
```
so, i would move this translation one level above. `r/private` should be fairly agnostic about how it's being used. let's just have `newsletter` be a bool.

nramadas commented on lib/api.es6.js:
```diff
@@ -50,6 +50,85 @@ const login = (apiOptions, username, pass) => {
+            obj.errorType = obj.status;
+          }
+
+          // the error response for login is different than most of the rest of
```
"the error response for login" -> "the error response for register"

nramadas commented on lib/api.es6.js:
```diff
@@ -50,6 +50,85 @@ const login = (apiOptions, username, pass) => {
+
+          // the error response for login is different than most of the rest of
+          // the api
+          if (res.body.json && res.body.json.errors && res.body.json.errors.length) {
```
if this check is shared, let's pull it out into a function. It would clean up the code:

``` es6
const errors = parseErrors(res);
if (errors.length) {
  return x(errors);
}
```

nramadas commented on lib/api.es6.js:
```diff
@@ -50,6 +50,85 @@ const login = (apiOptions, username, pass) => {
+          }
+
+          // No errors here
+          const cookies = (res.header['set-cookie'] || []).map(c => {
```
same here. i think parsing cookies is shared with login, so let's break it out.

``` es6
const cookies = parseCookies(res);
```

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
 };
 
+const register = (apiOptions, username, password, email, newsletter) => {
+  return new Promise((r, x) => {
```
I know it's a tad bit more verbose, but it'd be nice for these to be `(resolve, reject)`. It's slightly less overhead for the initiated and a lot friendlier for any ES6 newbies stumbling upon this code.

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
+      data.newsletter_subscribe = true;
+
+      if (!email) {
+        x('NEWSLETTER_NO_EMAIL');
```
Is there a reason some of these rejection messages seem like "token" type messages while some seem more colloquial?

jamesmaa replied:
I was caught in between several conflicting protocols for error messages that I didn't put much thinking time into resolving. 

I was using the error strings from the login function but then I found out that errors strings could also come from making calls to `r2`, which are like the "token" type of messages. Morever, these errors strings aren't properly handled on the frontend. For login, any type of error gives an `Sorry that's not the right password` error, whereas I wanted more granular and helpful error messages for register. 

Straddling between consistency versus helpfulness, I hot-patched this frankenstein together.

I think what would really help is if I knew where I would consolidate error messages, then they won't be thrown everywhere

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
+    };
+
+    try {
+      const timeout = 20000;
```
unused?

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
+        .type('form')
+        .send(data)
+        .timeout(10000)
+        .end((err, res) => {
```
:no_bell: I believe superagent conforms to the Promise interface, which means you can use `then` and `catch` here (if you want to). This looks fine otherwise.

nramadas replied:
i don't think there is a `catch`, just a `then`. for what it's worth, i actually prefer using `.end` with superagent so that it doesn't make it look like superagent uses real promises.

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
+          x('UNKNOWN_ERROR');
+       });
+    } catch (err) {
+      return x('Invalid registration information.');
```
Is this valuable information? What places are you worried about exceptions in the block above? Can we give it more context perhaps?

nramadas replied:
i think this was just copied from the errors that the old `snoode` used to provide. we did the same thing when we made the 'login' and 'refresh' endpoints.

If we're okay with changing these, let's go ahead and fix all of our error messages. Rejecting with a string is not very helpful. It'll be nice if our errors had more context, and if they used api-clients' actual error objects.

jamesmaa replied:
Actually you bring up a good point. I'm not worried about this anymore. 

I was debugging an opaque error, and thrown this in to surface the error. I'm not sure if the try catch block is needed anymore

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
+
+// Parses the error message for login and register functions
+const parseError = (res) => {
+  if (res.body.json && res.body.json.errors && res.body.json.errors.length) {
```
You don't really need `res.body.json.errors.length` here. If there were 0 items in the array it would still do what your fallback on line 131 is doing.

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
+
+// Parses the cookies for the login and register functions
+const parseCookies = (res) => {
+  return (res.header['set-cookie'] || []).map(c => {
```
bit of a nit, you can single line this:

``` js
const parseCookies = res => {
  return (res.header['set-cookie'] || []).map(c => c.split(';')[0])
}
```

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
+          const cookies = parseCookies(res);
+
+          const redditSessionExists = sessionExists(res);
+          if (redditSessionExists) {
```
Is there a bug here?
The function `sessionExists` is defined using `indexOf`, which will return -1 if there's not a match, which is actually truthy so you're going to be getting false positives here.

jamesmaa replied:
Good catch, that means the login logic is broken too
![image](https://cloud.githubusercontent.com/assets/2382537/17600189/39579044-5fb7-11e6-89bb-0affaed9f905.png)

nramadas commented on lib/api.es6.js:
```diff
@@ -50,6 +47,102 @@ const login = (apiOptions, username, pass) => {
+
+// Parses set-cookie header to see if the session exists
+const sessionExists = (res) => {
+  return res.header['set-cookie'].join('').indexOf('reddit_session');
```
to fix the bug phil pointed out above, you can use the `~` operator:

``` es6
return ~res.header['set-cookie'].join('').indexOf('reddit_session');
```

or, since that looks a bit ridiculous :), you can just use a `> -1` check;

phil303 replied:
strong preference for the latter

nramadas replied:
but think of the characters you could save!! /s

the `~` operator is a pretty odd one. it takes a value `n` and returns `-(n + 1)`. So, `-1` becomes `0`, `-2` becomes `1`, `-3` becomes `2`, etc. But, I can't see any possible use for it outside of turning `-1`s from an indexOf check into a falsey value. I'm pretty sure it was created _specifically_ for that purpose, just so that you could write `if (~foo.indexOf(bar))` easier.

phil303 replied:
Na, it's a bitwise operator that flips all bits in the thing it's operating on. In the indexOf situation, it's basically a convienent hack - `~-1` (32 1's in binary) becomes 0 which is falsey.

jamesmaa commented:
UPDATE: 
Responded to all comments
- removed try/catch surrounding superagent
- decided to go for the token type messages to parse in the frontend (PR [here](https://github.com/reddit/reddit-mobile/pull/672)). Non token types are for developer use
- Squashed all the other bugs

phil303 commented on lib/api.es6.js:
```diff
@@ -50,6 +46,95 @@ const login = (apiOptions, username, pass) => {
+}
+
+// Parses the error message for login and register functions
+const parseError = (res) => {
```
i didn't catch this before but this should probably be `parseErrors`

phil303 commented:
@jamesmaa in regards to your original question. As far as I know, only mweb 2X has tests. I think the other repos have been in a lot of flux and everyone's been pretty heads down on finishing up mweb so tests haven't been prioritized. But if you're feeling ambitious, tests are always, always welcome.

QA right now means linking it up to the mweb repo and testing it locally. Since we're not in prod yet, it doesn't make sense to have anything more than that just yet. I was out last week but I believe we're getting closer to getting a staging environment set up.

As far as this patch goes, code looks good. I'm not in the original set of reviewers but I'm gonna :fish: ya anyway :)

jamesmaa commented:
Thanks @phil303 for the thorough review! I added two more commits.
1. Because of [this](https://github.com/reddit/reddit-mobile/pull/672#discussion_r74502859) comment I moved the error parsing logic here. 
2. I refactored the login endpoint to be consistent to register. Now they look pretty much identical aside from the endpoint and data.

jamesmaa commented:
💇

nramadas commented:
:fish:

Commit:
v0.2.0
